tests/testthat/test-content.R

```diff
@@ -180,3 +180,48 @@ without_internet({
     )
   })
 })
+
+with_mock_api({
+  test_that("content_render() calls the correct endpoint, returns task on success", {
+    client <- Connect$new(server = "https://connect.example", api_key = "not-a-key")
+    x <- content_item(client, "951bf3ad-82d0-4bca-bba8-9b27e35c49fa")
+    render_task <- content_render(x)
+    expect_equal(render_task$task[["id"]], "v9XYo7OKkAQJPraI")
+    expect_equal(render_task$connect, client)
+  })
+
+  test_that("content_render() raises an error when called on interactive content", {
+    client <- Connect$new(server = "http://connect.example", api_key = "not-a-key")
+    x <- content_item(client, "8f37d6e0-3395-4a2c-aa6a-d7f2fe1babd0")
+    expect_error(content_render(x), "Render not supported for application mode: shiny. Did you mean content_restart()?", fixed = TRUE)
+  })
+
+  test_that("content_render returns an error when called on incorrect class", {
+
+  })
+})
+
+with_mock_api({
+  test_that("content_restart() calls the correct endpoint", {
+    client <- Connect$new(server = "https://connect.example", api_key = "not-a-key")
+    x <- content_item(client, "8f37d6e0-3395-4a2c-aa6a-d7f2fe1babd0")
+    expect_PATCH(content_restart(x)) # No returned value, so we just test that it succeeded.
+    # TODO: Understand if it's possible to set expectations when using a randomized env var.
+  })
+
+  test_that("content_restart() raises an error when called on interactive content", {
+    client <- Connect$new(server = "http://connect.example", api_key = "not-a-key")
+    x <- content_item(client, "951bf3ad-82d0-4bca-bba8-9b27e35c49fa")
+    expect_error(content_restart(x), "Restart not supported for application mode: quarto-static. Did you mean content_render()?", fixed = TRUE)
+  })
+})
+
+with_mock_api({
+  test_that("content$default_variant gets the default variant", {
+    client <- Connect$new(server = "http://connect.example", api_key = "not-a-key")
+    x <- content_item(client, "951bf3ad-82d0-4bca-bba8-9b27e35c49fa")
+    v <- x$default_variant # TODO Can't figure out why I can't use `expect_GET`
```

Comment:
What's the error you're getting with `expect_GET`?

Reply:
Something like

```
> without_internet({
+   with_mock_api({
+     test_that("content$default_variant gets the default variant", {
+       client <- Connect$new(server = "[https://connect.example](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)", api_key = "not-a-key")
+       x <- content_item(client, "951bf3ad")
+       # v <- x$default_variant # TODO Can't figure out why I can't use `expect_GET`
+       expect_GET(x$default_variant, url="[https://connect.example/__api__/applications/951bf3ad/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
+       # expect_identical(v$key, "WrEKKa77")
+     })
+   })
+ })
Defining Connect with server: [https://connect.example](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)
-- Failure: content$default_variant gets the default variant -------------------
`object` did not throw the expected error.
Backtrace:
    x
 1. \-httptest::expect_GET(x$default_variant, url = "[https://connect.example/__api__/applications/951bf3ad/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
 2.   \-httptest:::expect_mock_request(object, "GET ", url, " ", ...)
 3.     \-request_happened()(...)

-- Warning: content$default_variant gets the default variant -------------------
Arguments in `...` must be used.
x Problematic arguments:
* fixed = fixed
* ignore.case = ignore.case
* perl = perl
* useBytes = useBytes
i Did you misspell an argument name?
Backtrace:
     x
  1. +-httptest::expect_GET(x$default_variant, url = "[https://connect.example/__api__/applications/951bf3ad/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
  2. | \-httptest:::expect_mock_request(object, "GET ", url, " ", ...)
  3. |   \-request_happened()(...)
  4. |     \-testthat:::expect_condition_matching(...)
  5. |       \-rlang (local) `<fn>`()
  6. |         \-rlang:::check_dots(env, error, action, call)
  7. |           \-rlang:::action_dots(...)
  8. |             +-rlang (local) try_dots(...)
  9. |             | \-rlang::try_fetch(expr, error = error)
 10. |             |   +-base::tryCatch(...)
 11. |             |   | \-base (local) tryCatchList(expr, classes, parentenv, handlers)
 12. |             |   |   \-base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 13. |             |   |     \-base (local) doTryCatch(return(expr), name, parentenv, handler)
 14. |             |   \-base::withCallingHandlers(...)
 15. |             \-rlang (local) action(...)
 16. |               \-rlang:::signal_abort(cnd, .file)
 17. |                 \-base::signalCondition(cnd)
 18. \-rlang (local) `<fn>`(`<rlb_rr__>`)
 19.   \-handlers[[1L]](cnd)

Error:
! Test failed
Hide Traceback
     x
  1. +-httptest::without_internet(...)
  2. | \-base::eval.parent(expr)
  3. |   \-base::eval(expr, p)
  4. +-httptest::with_mock_api(...)
  5. | \-base::eval.parent(expr)
  6. |   \-base::eval(expr, p)
  7. +-testthat::test_that(...)
  8. | \-withr (local) `<fn>`()
  9. \-reporter$stop_if_needed()
 10.   \-rlang::abort("Test failed", call = NULL)
```

Reply:
I'm sure I'm doing something obviously wrong; I just can't see what yet.

Reply:
The error was the same with and without it; that was just the latest iteration. I also tried with the full and shortened GUID.

[P.S. The URLs in the console output are normal-looking URLs, not the weird vscode terminal markdown entities you see here]

```
> with_mock_api({
+   test_that("content$default_variant gets the default variant", {
+     client <- Connect$new(server = "[http://connect.example](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)", api_key = "not-a-key")
+     x <- content_item(client, "951bf3ad-82d0-4bca-bba8-9b27e35c49fa")
+     # v <- x$default_variant # TODO Can't figure out why I can't use `expect_GET`
+     expect_GET(x$default_variant, url="[https://connect.example/__api__/applications/951bf3ad/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
+     # expect_GET(x$default_variant, url="[https://connect.example/__api__/applications/951bf3ad-82d0-4bca-bba8-9b27e35c49fa/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
+     # expect_identical(v$key, "WrEKKa77")
+   })
+ })
Defining Connect with server: [http://connect.example](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)
-- Failure: content$default_variant gets the default variant -------------------
`object` did not throw the expected error.
Backtrace:
    x
 1. \-httptest::expect_GET(x$default_variant, url = "[https://connect.example/__api__/applications/951bf3ad/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
 2.   \-httptest:::expect_mock_request(object, "GET ", url, " ", ...)
 3.     \-request_happened()(...)

-- Warning: content$default_variant gets the default variant -------------------
Arguments in `...` must be used.
x Problematic arguments:
* fixed = fixed
* ignore.case = ignore.case
* perl = perl
* useBytes = useBytes
i Did you misspell an argument name?
Backtrace:
     x
  1. +-httptest::expect_GET(x$default_variant, url = "[https://connect.example/__api__/applications/951bf3ad/variants](vscode-file://vscode-app/Applications/Positron.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html#)")
  2. | \-httptest:::expect_mock_request(object, "GET ", url, " ", ...)
  3. |   \-request_happened()(...)
  4. |     \-testthat:::expect_condition_matching(...)
  5. |       \-rlang (local) `<fn>`()
  6. |         \-rlang:::check_dots(env, error, action, call)
  7. |           \-rlang:::action_dots(...)
  8. |             +-rlang (local) try_dots(...)
  9. |             | \-rlang::try_fetch(expr, error = error)
 10. |             |   +-base::tryCatch(...)
 11. |             |   | \-base (local) tryCatchList(expr, classes, parentenv, handlers)
 12. |             |   |   \-base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 13. |             |   |     \-base (local) doTryCatch(return(expr), name, parentenv, handler)
 14. |             |   \-base::withCallingHandlers(...)
 15. |             \-rlang (local) action(...)
 16. |               \-rlang:::signal_abort(cnd, .file)
 17. |                 \-base::signalCondition(cnd)
 18. \-rlang (local) `<fn>`(`<rlb_rr__>`)
 19.   \-handlers[[1L]](cnd)

Error:
! Test failed
Hide Traceback
    x
 1. +-httptest::with_mock_api(...)
 2. | \-base::eval.parent(expr)
 3. |   \-base::eval(expr, p)
 4. +-testthat::test_that(...)
 5. | \-withr (local) `<fn>`()
 6. \-reporter$stop_if_needed()
 7.   \-rlang::abort("Test failed", call = NULL)
```

Reply:
Talked to @nealrichardson about this. In R `httptest` the expect functions are designed to be used when mocks _aren't_ present. So, we're all good!